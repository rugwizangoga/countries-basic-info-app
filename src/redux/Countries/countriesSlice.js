import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

// Define the asynchronous thunk
export const fetchCountries = createAsyncThunk(
  'countries/fetchCountries',
  async () => {
    const response = await axios.get('https://restcountries.com/v3.1/all');
    return response.data;
  },
);

// Create the slice
const countriesSlice = createSlice({
  name: 'countries',
  initialState: {
    countries: [],
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchCountries.pending, (state) => ({
        ...state,
        loading: true,
        error: null,
      }))
      .addCase(fetchCountries.fulfilled, (state, action) => ({
        ...state,
        loading: false,
        countries: action.payload,
      }))
      .addCase(fetchCountries.rejected, (state, action) => ({
        ...state,
        loading: false,
        error: action.error.message,
      }));
  },
});

// Export the reducer
export default countriesSlice.reducer;
