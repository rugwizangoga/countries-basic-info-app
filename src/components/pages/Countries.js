import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchCountries } from '../../redux/Countries/countriesSlice';
import Country from './Country';

const Countries = () => {
  const countriesList = useSelector((store) => store.countries.countries);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCountries());
  }, []);

  const sortedCountriesList = [...countriesList]
    .sort((a, b) => a.name.common.localeCompare(b.name.common));

  return (
    <>
      <div>
        {sortedCountriesList.map((country) => (
          <div key={country.name.common}>
            <img
              src={country.flags.png}
              alt="flag"
              className="flag-img"
            />
            {country.name.common}
          </div>
        ))}
      </div>
      <Country />
    </>
  );
};
export default Countries;
