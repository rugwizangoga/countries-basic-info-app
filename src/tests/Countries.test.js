import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import Countries from '../components/pages/countries';
import store from '../redux/store'; // Replace with your actual Redux store import

test('renders Country component correctly', () => {
  const { container } = render(
    <Provider store={store}>
      <Countries />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
