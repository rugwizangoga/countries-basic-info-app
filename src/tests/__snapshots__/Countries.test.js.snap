// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Country component correctly 1`] = `
<div>
  <div
    class="app"
  >
    <img
      alt="World map"
      class="wmap"
      src="wmap.png"
    />
    <h1>
      Number of official languages per country
    </h1>
    <div
      class="info"
    />
  </div>
</div>
`;
